### PR TITLE
update mypy to latest (WIP)

### PR DESCRIPTION
**Pull request by stillmatic**
**Description**
- Updates the mypy version from 0.600 to 0.910
- Removes the `protobuf` stubs since latest mypy includes them
- Removes `six` usages as Python 2 support is deprecated
- Fixes issues as they arise

**Motivation and Context**
- Allows us to use modern mypy versions with onnx 
- Previous works and more context: #3211

### CODE REVIEW

**Review by jcwchen — COMMENTED** [review]
Hi @stillmatic,
I am happy to see this coming in ONNX since the current mypy typecheck is quite old. Thank you for the effort! Is there anything I can do to help you accomplish this PR?

**Comment by stillmatic** [comment]
@jcwchen Thanks for taking a look (and reminding me!) 

I rebased this PR off of the main branch and it looks like tests are now passing. Please LMK what else would be helpful to do!

**Review by jcwchen — COMMENTED** [review]
Thank you for fixing the CI failures! IIUC, six is not needed anymore, right? If so, please remove six dependency in README.md, requirements.txt, requirements-dev.txt, requirements-release.txt.

**Comment by stillmatic** [comment]
Thanks for the catch - removed all references to six

File: onnx/backend/test/runner/__init__.py [hunk 1]
```diff
@@ -168,7 +167,7 @@ def tests(self):  # type: () -> Type[unittest.TestCase]
             onnx_backend_tests = BackendTest(backend).tests
         '''
         tests = self._get_test_case('OnnxBackendTest')
-        for items_map in sorted(self._filtered_test_items.values()):
+        for items_map in sorted(self._filtered_test_items.values()):  # type:ignore
```
Comment (jcwchen) [inline onnx/backend/test/runner/__init__.py:170]: nit: sync all `# type: ignore` with the same space (there are 4 places)
```suggestion
        for items_map in sorted(self._filtered_test_items.values()):  # type: ignore
```
Thank you for the update! mypy upgrade seems OK for ONNXRUntime as well. I hope we can forward this PR soon.

One question: Do you know why the same function here need to be ignored for the updated mypy?

**Comment by jcwchen** [comment]
Superseded by https://github.com/onnx/onnx/pull/3939. Close this one.

**Comment by jcwchen** [comment]
Now ONNX has removed six https://github.com/onnx/onnx/pull/3926 and used a newer mypy version (0.760) https://github.com/onnx/onnx/pull/3926. Reopen it since actually this one is for upgrading to the latest mypy. (Or feel free to submit a new clean PR for it). Thank you for all the contribution!

**Comment by stillmatic** [comment]
thanks @jcwchen  -- let's close this and track followups separately. most of the valuable pieces from this PR have been broken out now :)

**Comment by garymm** [comment]
Is there an issue or open PR for updating mypy?

Using the version that's currently in setup.py (0.760) I see this error:
```
➜  python setup.py typecheck
running typecheck
onnx/__init__.py:30: error: syntax error in type comment
```

I think this may be a mypy bug that's fixed in a later version.

**Comment by jcwchen** [comment]
> Is there an issue or open PR for updating mypy?

There isn't yet. @stillmatic Do you still have bandwidth to propose a new PR for it? Thank you!

> I think this may be a mypy bug that's fixed in a later version.

@garymm Are you talking about the error in this PR https://github.com/onnx/onnx/pull/3962 ? If so, yes I think latest mypy should support Python type annotation according to the document: https://mypy.readthedocs.io/en/stable/type_inference_and_annotations.html.

**Comment by stillmatic** [comment]
I will open an issue to track the work but likely won't have much more bandwidth for implementation: https://github.com/onnx/onnx/issues/3964

Hmm - is the type check path not part of the CI tests? I also get a different error when running that command.

```
~/Development/onnx  main ✔                                                                     1h6m  ⍉
▶ ~/Development/onnx  main ✔                                                                      1h6m
▶ mypy --version
mypy 0.760
(onnx)
~/Development/onnx  main ✔                                                                      1h6m
▶ python setup.py typecheck
/Users/hua/.pyenv/versions/onnx/lib/python3.8/site-packages/setuptools/installer.py:27: SetuptoolsDeprecationWarning: setuptools.installer is deprecated. Requirements should be satisfied by a PEP 517 installer.
  warnings.warn(
running typecheck
tools/protoc-gen-mypy.py:122: error: Missing type parameters for generic type "Generator"
Found 1 error in 1 file (checked 245 source files)
(onnx)
```

**Comment by garymm** [comment]
@jcwchen 
>  Are you talking about the error in this PR #3962 ?

No, this is on tip of main with no changes.

@stillmatic I'm using python 3.9.

I guess `python setup.py typecheck` fails in multiple ways on tip of main, depending on python version used. Not sure why this isn't caught by CI checks.

@jcwchen should I open another issue to track the current failures?

**Comment by jcwchen** [comment]
> @jcwchen should I open another issue to track the current failures?

Yes, that would be great. Thanks! Actually current CIs are only using Python 3.6 and 3.7 (no 3.8 and 3.9) for typecheck. I tried to run 3.8 and 3.9 in CIs and bumped into similar errors to yours. We should add Python 3.8 and 3.9 in AzurePipeline as well to cover this. I will pick it up. Thank you both for catching this!

**Pull request closed without merging**